polar-core/src/formatting.rs

```diff
@@ -57,36 +54,27 @@ impl Trace {
     }
 }
 
-/// Traverse a [`Source`](../types/struct.Source.html) line by line until `offset` is reached,
-/// and return the source line containing the `offset` character as well as `num_lines` lines
-/// above and below it.
+/// Traverse a [`Source`](../types/struct.Source.html) line by line until `offset` is reached and
+/// return the source line containing the `offset` character as well as `num_lines` lines above and
+/// below it.
 // @TODO: Can we have the caret under the whole range of the expression instead of just the beginning.
 pub fn source_lines(source: &Source, offset: usize, num_lines: usize) -> String {
-    // Sliding window of lines: current line + indicator + additional context above + below.
-    let max_lines = num_lines * 2 + 2;
-    let push_line = |lines: &mut Vec<String>, line: String| {
-        if lines.len() == max_lines {
-            lines.remove(0);
-        }
-        lines.push(line);
-    };
-    let mut index = 0;
-    let mut lines = Vec::new();
-    let mut target = None;
+    let (target_line, target_column) = loc_to_pos(&source.src, offset);
+    // Skip everything up to the first line of requested context (`target_line - num_lines`), but
+    // don't overflow if `num_lines > target_line`.
+    let skipped_lines = target_line.saturating_sub(num_lines);
+    let lines = source.src.lines().skip(skipped_lines);
+    // Take window of lines comprising current line + `num_lines` context above & below.
+    let lines = lines.take(1 + num_lines * 2).enumerate();
+    // Format each line with its line number.
+    let mut lines: Vec<_> = lines
+        .map(|(i, line)| format!("{:03}: {}", i + skipped_lines + 1, line))
+        .collect();
+    // Calculate length of line number prefix.
     let prefix_len = "123: ".len();
-    for (lineno, line) in source.src.lines().enumerate() {
-        push_line(&mut lines, format!("{:03}: {}", lineno + 1, line));
-        let end = index + line.len() + 1; // Adding one to account for new line byte.
-        if target.is_none() && end >= offset {
-            target = Some(lineno);
-            let spaces = " ".repeat(offset - index + prefix_len);
-            push_line(&mut lines, format!("{}^", spaces));
-        }
-        index = end;
-        if target.is_some() && lineno == target.unwrap() + num_lines {
-            break;
-        }
-    }
+    // Insert 'indicator' line pointing at `target_column`.
+    let indicator_line = format!("{}^", " ".repeat(prefix_len + target_column));
+    lines.insert(target_line - skipped_lines + 1, indicator_line);
```

Comment:
I might be mistaken, but could this be simplified to `lines.insert(num_context_lines + 1, indicator_line)` ?

Reply:
indexing the shorter `Vec` which is only `(num_context_lines * 2) + 1` items long by using the indices of the wider `lines` Vec was confusing

Reply:
> I might be mistaken, but could this be simplified to `lines.insert(num_context_lines + 1, indicator_line)` ?

Is `num_context_lines` the third function argument (now called `context_lines`)? If so, I don't think that holds true for the case where you say you'd like 2 lines of context on either side of the target line but the source text only contains 1 line. You'd be trying to insert at position `3` into a `Vec` of length `1`.

Reply:
> indexing the shorter `Vec` which is only `(num_context_lines * 2) + 1` items long by using the indices of the wider `lines` Vec was confusing

But it might not actually be `(context_lines * 2) + 1` items long — that's just the _maximum_ size the window can be

Reply:
ah yes! of course.

Reply:
(to be clear: I definitely agree that it's confusing 😆)